### PR TITLE
Fix U2F registration and signing by replacing jQuery requests by fetch

### LINKED ISSUE #11228: U2F doesn't work on gitea/gitea:latest docker image
- Gitea version (or commit ref): docker: gitea/gitea:latest
- Git version: n/a
- Client Operating system: Windows 10
- Client browser: Firefox and Edge
- Host OS: CentOS 7
- Host Docker version: Docker CE version 19.03.8
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [x] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes (provide example URL)
  - [ ] No
  - [ ] Not relevant

## Description
When I upgrade my docker container to the newest gitea release (gitea/gitea:latest) the yubikey authentication attempt causes the error attached. When I downgrade the gitea to 1.11.4 (gitea/gitea:1.11.4) the yubikey authentication works without issues.

## Screenshots
gitea_u2f_error

### PR DESCRIPTION
**Pull request by jonasfranz**
Fix #11228

### CODE REVIEW

**Comment by silverwind** [comment]
👍 to moving it to a separate file if possible.

If we expect this to work in IE11 we should include a [fetch polyfill](https://github.com/github/fetch).

**Comment by jonasfranz** [comment]
@6543 I'm currently working on a rewrite to support WebAuthn. Do you think it's worth it to rewrite this now? You can have a look on the new implementation here: https://github.com/go-gitea/gitea/compare/master...jonasfranz:feature/webauthn?expand=1

I saw that @e3b0c442 already started working on webauthn while I'm writing this comment :joy: .

**Comment by silverwind** [comment]
JS looking good. 

I wonder if we can switch the vendored `u2f-api` to the [npm version](https://www.npmjs.com/package/u2f-api)? Documentation points to it being from this [fork](https://github.com/go-gitea/u2f-api), is there a specific reason?

**Comment by e3b0c442** [comment]
Don't let the WebAuthn stuff keep you from doing this, it looks like this is needed to fix broken functionality and I'm not likely to have time to dive hard back into WebAuthn until my classes are over in another 6 weeks.

**Comment by e3b0c442** [comment]
> @6543 I'm currently working on a rewrite to support WebAuthn. Do you think it's worth it to rewrite this now? You can have a look on the new implementation here: https://github.com/go-gitea/gitea/compare/master...jonasfranz:feature/webauthn?expand=1
> 
> I saw that @e3b0c442 already started working on webauthn while I'm writing this comment  .

I have, and I'm hoping to get back to that in about 6 weeks when my classes are over for the summer, that said, if you're in a position to run with it go right ahead, won't hurt my feelings.

**Comment by jonasfranz** [comment]
@silverwind the npm version is not compatible with the current implementation. I would also love to use the npm version instead. But that might not be worth the effort since it will be replaced by webauthn in some point in time anyway.

**Review by silverwind — APPROVED** [review]

**Comment by silverwind** [comment]
I wonder why `package-lock.json` hasn't changed. Can you run a `npm install --package-lock --package-lock-only` and commit the file if it changed?

**Comment by jonasfranz** [comment]
@silverwind package-lock.json didn't change since we `swagger-client` has `cross-fetch` as dependency which depends on the added package. The proposed commands also didn't change the lock file.

**Comment by silverwind** [comment]
Right, I see `whatwg-fetch` in there so everything's fine.

**Comment by 6543** [comment]
fix do not work for me :(

**Comment by jonasfranz** [comment]
@6543 can you give me more details? Are getting an error message for example in the browser console?

File: web_src/js/features/webauthn.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    const signature = await u2fApi.sign(appId, challenge, registeredKeys, 30);
+    u2fSigned(signature);
+  } catch (e) {
+    if (e === undefined || e.metaData === undefined) {
+      u2fError(1);
+      return;
+    }
+    u2fError(e.metaData.code);
+  }
+}
```
Comment (6543) [inline web_src/js/features/webauthn.js:26]:
```suggestion
}

```

**Comment by 6543** [comment]
> @6543 can you give me more details? Are you getting an error message for example in the browser console?

` Could not read your security key.
Your browser does not support U2F security keys. `

System: linux 5.6.8; 76.0 (64-bit)

**Comment by 6543** [comment]
same on google chrome (Version 81.0.4044.138 (Official Build) (64-bit))

**Comment by zeripath** [comment]
@6543 are you running on a localhost? Is it possible that you have to have ssl set-up?

**Comment by 6543** [comment]
> @6543 are you running on a localhost? Is it possible that you have to have ssl set-up?

with self signed cert: no change :(

**Comment by CirnoT** [comment]
@6543 Make sure you type anything in nickname field, empty one is not accepted and returns `409 Conflict`.

**Comment by 6543** [comment]
@CirnoT if it would be that easy ...

**Comment by CirnoT** [comment]
Does it still show `Your browser does not support U2F security keys`? For me it simply shows `Could not read your security key` without anything about browser, but I assume that is because I've tested it on Windows 10 1909, so that's expected.

**Comment by CirnoT** [comment]
Interestingly the code seems to be there to handle 409 - 

https://github.com/go-gitea/gitea/pull/11311/files#diff-0fc7c09bb090d93aa9e9ce1e52c5a3a1R112

It doesn't seem to work however, it just shows `Your browser does not support U2F security keys`.

**Comment by 6543** [comment]
ok message is different with ssl now `Could not read your security key. `

**Comment by jonasfranz** [comment]
I close this PR since @CirnoT fixed the problem with PR #11379.

**Comment by silverwind** [comment]
Should still move the u2f parts to its own file later.

**Pull request closed without merging**